pkg/apis/eventing/v1alpha1/cluster_provisioner_types.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"encoding/json"
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/webhook"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterProvisioner provisions.
+type ClusterProvisioner struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterProvisionerSpec   `json:"spec"`
+	Status ClusterProvisionerStatus `json:"status"`
+}
+
+// Check that ClusterProvisioner can be validated and can be defaulted.
+var _ apis.Validatable = (*ClusterProvisioner)(nil)
+var _ apis.Defaultable = (*ClusterProvisioner)(nil)
+var _ runtime.Object = (*ClusterProvisioner)(nil)
+var _ webhook.GenericCRD = (*ClusterProvisioner)(nil)
+
+// ClusterProvisionerSpec is the spec for a ClusterProvisioner resource.
+type ClusterProvisionerSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Type is the type of the resource to be provisioned.
+	// +required
+	Type runtime.TypeMeta `json:"type"`
```

Comment:
I think this should probably be `metav1.TypeMeta` since it's the API object used by all other types. I'm not sure what `runtime.TypeMeta` is for. Their definitions are identical.